### ISSUE TITLE
Allow to specify error handling for watch()

### ISSUE THREAD

**Issue by martinbonnin** [apollographql/apollo-kotlin#3786]
Currently, `watch()` ignores all errors (not only cache) which is different from 2.x. 

I'm not sure what the different use cases are but getting an error for the initial request seems like an important thing. To allow for maximum configurability, what about adding a flag to watch:

```kotlin
enum class WatchErrorHandling {
  EMIT_CACHE_ERRORS,
  EMIT_NETWORK_ERRORS,
  EMIT_CACHE_AND_NETWORK_ERRORS,
  IGNORE_ERRORS
}

fun ApolloQueryCall<D>.watch(errorHandling: WatchErrorHandling)
```

This can be complexified that we have both `fetchPolicy` and `refetchPolicy` at the moment so we might want to have separate error handling there but a possible solution could be to leverage coroutines instead:

```kotlin
// The initial query will ignore cache errors and rethrow other errors
apolloClient.query(query).toFlow()
 .catch { if (it !is CacheMissException) throw it}
 .onCompletion {
  emitAll(apolloClient.query.watch())
}
```

**Comment by peterplamenovpenchev** [comment]
Hey all, we are currently migrating to Apollo 3.0 & it is a lot of fun.
We stumbled upon that discrepancy as well.

Is there any visibility when will this be available? I guess in 3.0.1 ?

**Comment by BoD** [comment]
In the end we went for a simpler API: `watch(fetchThrows: Boolean = false, refetchThrows: Boolean = false)` and also added a lower level `watch(data: D?, retryWhen: suspend (cause: Throwable, attempt: Long) -> Boolean = { _, _ -> true })` API that can be used in more advanced scenarios.

This will be in the next version - in the meantime can be tried out with the [snapshots](https://github.com/apollographql/apollo-kotlin#snapshots).

**Comment by martinbonnin** [comment]
Fixed in 3.1.0
